Ocaml benchmarks

gasche opened the pull request:
This is a bit of a monster PR as I am commuting out, sorry.

A lot of this is scripts I use to produce the LaTeX tables and plots automatically. But the rest is also interesting:
- zipper implementation of replay non-determinism, in OCaml and SML (including array/vector versions to remove the (List.nth) inherent in choose+replay)
- rename the SML files to follow the OCaml scheme (to have a unified description in the benchmark sections, if we decide to show both families of numbers)

The PR contains the `result.data` that I generated on my machine, which suffices to reproduce the `.tex` parts that I committed in the paper -- and change them if we tune the automated output again, etc.

gasche committed:
ocaml: use Obj.t as universal type

gasche committed:
minor clarification to thermometers.ml

gasche committed:
rename thermometers into thermometers_optimized

This code uses the nondet-specific optimizations of Section 6.2

gasche committed:
add a generic (non-optimized) implementation of thermometers

This code is based on the generic SML version delim.sml.

gasche committed:
cleanup Makefile

gasche committed:
copy thermometers_generic into thermometers_refactored

The idea is to use this version as a working version to check various
simplifications to the code.

gasche committed:
fix the delimcc benchmark

gasche committed:
minor refactoring

gasche committed:
shift/reset: add unit tests

gasche committed:
replay: use a zipper (two stacks)

The replay-based implementation using a zipper is sensibly more
efficient than the one based on a `pos` reference within a branch
index list, and much closer to the full thermometers implementation.

I am also adding a version where `choose` takes an array rather than
a list, to save on indexing costs there as well.

gasche committed:
add a nesting-safe replay implementation

Adding the nesting stack makes the implementation closer to
thermometers again.

gasche committed:
SML benchmarks: factor (range 0 n) computation

gasche committed:
SML benchmarks: add replay implementations with zipper

gasche committed:
benchmark script: add a LaTeX-friendly output mode

gasche committed:
rename nqueens SML benchmarks

gasche committed:
fix the OCaml nqueens benchmark

gasche committed:
benchmark script: configurable upper-bound, awk-friendly output mode

gasche committed:
delimcc now works with OCaml 4.06

gasche committed:
improve benchmark scripts

gasche committed:
AWK scripts to extract LaTeX tables and plots from the data

gasche committed:
shell script to produce the data

gasche committed:
the data from my machine (for easy rebuild)

gasche committed:
add a script to produce final LaTeX data in latex/ subdir

gasche committed:
fix printing of multi-minutes times

gasche committed:
improve labels in {only,no}sml outputs

gasche committed:
ThermSML was missing from the result data

jkoppel commented on benchmarks/nqueens/ocaml/thermometers_refactored.ml:
```diff
@@ -49,10 +49,10 @@ module Control (A : Answer) : CONTROL with type ans = A.ans = struct
-       cur_fun := f';
-       record_stack := rec_stack';
-       replay_stack := rep_stack';
+     | Some (future, record, replay) ->
```
Cute, but the cur_fun contains both the future and the past.

jkoppel commented on benchmarks/nqueens/run_benchmark.rb:
```diff
@@ -15,13 +15,18 @@
+  TRIALS.times {
+    unless best > 60*1000*2 then
+      run = `timeout 10m sml #{file} #{n}`.lines[-1].to_i
+      if best = 0 || run < best then best = run end
```
Did you mean "if best == 0"?

jkoppel commented on benchmarks/nqueens/run_benchmark_time.rb:
```diff
@@ -19,20 +19,23 @@
+      #p s
+      m = s.match(/user\s+(\d+)m(\d+.\d+)s/)
+      run = (m[1].to_i * 60 + m[2].to_f) * 1000
+      if best = 0 || run < best then best = run end
```
Ditto above

jkoppel commented:
I skimmed over everything. It looks good...

...except for the "if best = 0" issue, which looks quite concerning. What's going on there?

gasche commented:
Indeed; I tried to get the "best out of N" instead of "average over N" that people find more stable for benchmarks (I'm no expert, but was told that the outliers are more slow-outliers, so "best" filters them out). Right now the code only takes the first (or last?) run, I suppose. I will rerun the benchmarks with the fixed code.

gasche committed:
fix bug in time-counting script spotted by Jimmy

The time-counting script were always using the last measurement,
instead of the best one, because (=) and (==) are different
operators...

jkoppel commented:
Glad we caught that.
